### PR TITLE
refactor: mark `--verbose` as deprecated

### LINKED ISSUE #1533: Deprecate the `--verbose` flag
As per the [ORAS diagnose experience proposal](https://github.com/oras-project/oras/blob/main/docs/proposals/diagnose-experience.md), we should mark the `--verbose` flag as deprecated for `oras pull`, `oras push`, `oras attach`, and `oras discover` commands.

- For `oras pull`, `oras push`, `oras attach` commands, we should output the detailed status output by default. As a result, the output with and without `--verbose` will be identical.
- For `oras discover`, we do not change the default output.

https://github.com/oras-project/oras/blob/5280c67fcef9e1029d539aef03370a31e5df839a/docs/proposals/diagnose-experience.md?plain=1#L72-L74

### PR DESCRIPTION
**Pull request by Wwwsylvia**
**What this PR does / why we need it**:
1. Mark the `--verbose` flag as deprecated and print verbose output by default for the following six commands. As a result, the `--verbose` flag will be hidden in the help doc, and there will be a warning message when `--verbose` is specified.
- `oras push`
- `oras pull`
- `oras attach`
- `oras cp`
- `oras manifest push`
- `oras blob push`

2. Add the corresponding E2E tests

The deprecation of `--verbose` for `oras discover` will be addressed with #1534.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1533
Prerequisite: #1553

---
Examples:
`oras pull` BEFORE:
image

`oras pull` AFTER:
image

`oras pull --verbose` BEFORE:
image

`oras pull --verbose` AFTER:
image

`oras pull --verbose=false` BEFORE:
image

`oras pull --verbose=false` AFTER:
image

`oras pull --help` BEFORE:
image

`oras pull --help` AFTER:
image

**Please check the following list**:
- [x]  Does the affected code have corresponding tests, e.g. unit test, E2E test?
- [ ]  Does this change require a documentation update?
- [ ]  Does this introduce breaking changes that would require an announcement or bumping the major version?
- [ ]  Do all new files have an appropriate license header?

### CODE REVIEW

**Comment by TerryHowe** [comment]
Maybe mark it deprecated in the Printer object too

**Comment by Wwwsylvia** [comment]
> Maybe mark it deprecated in the Printer object too

Currently it is still functioning, we can consider removing the whole verbose stuff in v1.4.0 or something?
From my understanding, the printer will be refactored by @wangxiaoxuan273 soon as part of handler refactoring.

File: cmd/oras/root/attach.go [hunk 1]
```diff
@@ -44,7 +44,7 @@ type attachOptions struct {
 
 	artifactType string
 	concurrency  int
-	verbose      bool
+	verbose      bool // deprecated
```
Comment (shizhMSFT) [inline cmd/oras/root/attach.go:47]: Should we mark the `verbose` as deprecated based on the golang documentation?

**Review by shizhMSFT — APPROVED** [review]
LGTM with suggestions

File: cmd/oras/root/attach.go (hunk 1 above)
Comment (Wwwsylvia) [reply cmd/oras/root/attach.go]: Updated.

**Review by shizhMSFT — APPROVED** [review]
LGTM

File: test/e2e/internal/utils/deprecation.go
Comment (qweeah) [inline test/e2e/internal/utils/deprecation.go:1]: Can you merge this file into `test/e2e/internal/testdata/feature/feature.go`? The deprecation message of the verbose flag is a known test data.

File: test/e2e/internal/utils/match/status.go [hunk 2]
```diff
@@ -77,6 +77,9 @@ func newStateMachine(cmd string) *stateMachine {
 		sm.addPath("Skipped")
 		sm.addPath("Exists")
 		sm.addPath("Mounted")
+	case "manifest", "blob": // for `manifest push` and `blob push`
```
Comment (qweeah) [inline test/e2e/internal/utils/match/status.go:80]: What if we are adding status tests for `oras manifest delete` or `oras manifest blob delete`?

Looks like we need to improve `NewStatusMatcher` to take the whole argument and do a best-effort prefix matching.

File: test/e2e/suite/command/attach.go [hunk 3]
```diff
@@ -42,9 +42,36 @@ var _ = Describe("ORAS beginners:", func() {
 
 		It("should show preview and help doc", func() {
 			out := ORAS("attach", "--help").MatchKeyWords(feature.Preview.Mark+" Attach", feature.Preview.Description, ExampleDesc).Exec()
+			// verbose flag should be hidden in help doc
```
Comment (qweeah) [inline test/e2e/suite/command/attach.go:45]: Is there a check to echo this comment?

File: test/e2e/suite/command/blob.go [hunk 4]
```diff
@@ -39,6 +40,31 @@ var _ = Describe("ORAS beginners:", func() {
...
+			})
+
+			It("should show deprecation message and should NOT print unnamed status output for --verbose=false", func() {
+				repo := fmt.Sprintf(repoFmt, "push", "test-verbose-false")
+				out := ORAS("blob", "push", RegistryRef(ZOTHost, repo, pushDigest), "-", "--size", strconv.Itoa(len(pushContent)), "--verbose=false").
+					WithInput(strings.NewReader(pushContent)).
+					MatchErrKeyWords(DeprecationMessageVerboseFlag).
+					Exec()
+				// should not print status output for unnamed blobs
+				gomega.Expect(out).ShouldNot(gbytes.Say("application/vnd.oci.image.layer.v1.tar"))
```
Comment (qweeah) [inline test/e2e/suite/command/blob.go:65]:
```suggestion
				gomega.Expect(out.Out).ShouldNot(gbytes.Say("application/vnd.oci.image.layer.v1.tar"))
```

File: test/e2e/suite/command/manifest.go [hunk 5]
```diff
@@ -55,6 +56,8 @@ var _ = Describe("ORAS beginners:", func() {
 			It("should show help doc with feature flags", func() {
 				out := ORAS("manifest", "push", "--help").MatchKeyWords(ExampleDesc).Exec()
 				gomega.Expect(out).Should(gbytes.Say("--distribution-spec string\\s+%s", regexp.QuoteMeta(feature.Preview.Mark)))
+				// verbose flag should be hidden in help doc
+				gomega.Expect(out).ShouldNot(gbytes.Say("--verbose"))
```
Comment (qweeah) [inline test/e2e/suite/command/manifest.go:60]: 1. out.Out should be used for matching
2. Suggest adding new spec for verbose flag checking. The reason is out.Out is not reset after matching preview feature mark. You will get false positive when `--verbose` is printed before `--distribution-spec` flag.

**Review by FeynmanZhou — COMMENTED** [review]
The examples align with the spec and look good to me. Thanks @Wwwsylvia

**Comment by FeynmanZhou** [comment]
@TerryHowe would you mind taking a look at this PR?

File: test/e2e/internal/utils/match/status.go (hunk 2 above)
Comment (Wwwsylvia) [reply test/e2e/internal/utils/match/status.go:80]: We can open an issue to track this.

File: test/e2e/suite/command/blob.go (hunk 4 above)
Comment (Wwwsylvia) [reply test/e2e/suite/command/blob.go:65]: Fixed by adding `.Out` after `Exec()`.

File: test/e2e/suite/command/manifest.go (hunk 5 above)
Comment (Wwwsylvia) [reply test/e2e/suite/command/manifest.go]: Updated.

File: test/e2e/internal/utils/deprecation.go
Comment (Wwwsylvia) [reply test/e2e/internal/utils/deprecation.go:1]: Done.

File: test/e2e/internal/utils/match/status.go (hunk 2 above)
Comment (Wwwsylvia) [reply test/e2e/internal/utils/match/status.go:80]: Put a TODO in the code and opened an issue: https://github.com/oras-project/oras/issues/1571

**Review by TerryHowe — APPROVED** [review]
/lgtm

**Pull request merged**